Add direct method of authenticating with Ecobee

tedpearson opened the pull request:
Two new functions:
ecobee.Authorize()
ecobee.SaveToken()

These new functions allow apps that depend on this library to authorize with the ecobee api in different ways:
- Outside of a go-ecobee client request
- Non-interactively

I specifically needed this to deploy a personal project I built via ansible. Ansible doesn't let you interact with commands, so I needed a way to get the pin, `pause` the ansible play, then run a separate step to get the first token using the code. (Project: https://github.com/tedpearson/ecobee2influx)

Happy to modify this PR to any desired code styles as needed.

Commit:
Add GetPin and SaveToken to ecobee.Client

This allows for use with apps installed via ansible workflows which don't allow interactive shells (e.g. hitting enter after entering pin on ecobee.com in a browser)

Commit:
change module path to this github repo

Commit:
simplify GetPin/SaveToken to not be receiver methods

Uncomplicates matters by not requiring access to an internal tokensource in the client struct.

Commit:
Revert "change module path to this github repo"

This reverts commit a385588c588c53ccf38fc0cd0cad4043927e60a9.

Commit:
document new methods

Commit:
Formating

rspier commented on ecobee/auth.go:
```diff
@@ -60,7 +60,7 @@ func (ts *tokenSource) save() error {
 }
 
-func (ts *tokenSource) firstAuth() error {
+func (ts *tokenSource) FirstAuth() (PinResponse, error) {
```
Please return a *PinResponse, so it can be `nil` in the error case.

tedpearson replied:
I am changing this as requested. Counter-argument anyways: callers should always check an error before using the normal return value. No need to use a pointer for a simple struct. Still new to Go, does that seem like a valid argument?

tedpearson replied:
fixed

rspier replied:
I'm not finding the reference I was looking for.  But I believe the rationale is to avoid the unnecessary memory allocation.  (An unnecessary micro-optimization in this case.)  More importantly, this is the common pattern.  Structs are almost always passed by reference.

rspier commented on ecobee/auth.go:
```diff
@@ -76,36 +76,30 @@ func (ts *tokenSource) firstAuth() error {
-
-	return ts.accessToken(r.Code)
-
+type PinResponse struct {
```
Please move this before the first use.  I know it's not required, but it makes things a little easier when scanning the code.

tedpearson replied:
fixed

rspier commented on ecobee/auth.go:
```diff
@@ -60,7 +60,7 @@ func (ts *tokenSource) save() error {
 }
 
-func (ts *tokenSource) firstAuth() error {
+func (ts *tokenSource) FirstAuth() (PinResponse, error) {
```
I think this can stay firstAuth.  The exported function is going to be GetPin, so there's no reason to export FirstAuth.  (Am I missing something?)

tedpearson replied:
fixed

rspier commented on ecobee/auth.go:
```diff
@@ -206,6 +208,17 @@ type Client struct {
+
+// SaveToken retreives a new token from ecobee and saves it to the auth cache
+// after a pin/code combination has been added by an ecobee user.
+func SaveToken(clientId string, authCache string, code string) error {
```
authCache => cacheFile for consistency with above.

tedpearson replied:
Why do you want consistency with newTokenSource() instead of consistency with NewClient()? Should NewClient be changed as well perhaps?

rspier replied:
I didn't notice it was in both places.    So yes.

`authCache string` isn't as self-documenting  as `cacheFile string`

tedpearson replied:
fixed all occurrences

rspier commented on ecobee/auth.go:
```diff
@@ -186,7 +180,15 @@ func (ts *tokenSource) Token() (*oauth2.Token, error) {
 			}
 		} else {
-			err := ts.firstAuth()
+			pinResponse, err := ts.FirstAuth()
```
The code you moved from firstAuth feels a little out of place here.

What do you think about renaming firstAuth to authorize or similar, and then keeping this prompt in firstAuth?

tedpearson replied:
I agree and this is probably the most expected comment. What you propose sounds fine.

tedpearson replied:
fixed

rspier replied:
thank you

rspier commented on ecobee/auth.go:
```diff
@@ -206,6 +208,17 @@ type Client struct {
+
+// GetPin retrieves an ecobee Pin and Code, allowing calling code to present them to the user
+// outside of the ecobee request context.
+func GetPin(clientId string) (PinResponse, error) {
```
Would Authorize be a better name for this?

tedpearson replied:
I'm not sure. This is technically just the first step of a 2-step process to authorize. The "Get" part of the name may be bad, since it's not really a getter, maybe LookupPin or AuthStepOne or something? /shrug

tedpearson replied:
Well, 3-step if you count entering the pin into the website...

tedpearson replied:
Went with Authorize for the name anyways. Fixed.

rspier replied:
I chose Authorize based on what the ecobee API calls it.  I agree it's not perfect.

https://www.ecobee.com/home/developer/api/documentation/v1/auth/auth-intro.shtml

rspier commented on ecobee/auth.go:
```diff
@@ -206,6 +208,17 @@ type Client struct {
+}
+
+// GetPin retrieves an ecobee Pin and Code, allowing calling code to present them to the user
+// outside of the ecobee request context.
```
Maybe add a sentence about how this is useful when you need to split Authorization from creation of the access token itself?

tedpearson replied:
👍

tedpearson replied:
fixed

rspier reviewed (commented):
Thank you!  I'm glad you're finding the library useful.  This PR looks good.  I left some minor comments, and I'm definitely interested in your opinion on them.

Thanks!

-R

Commit:
Code review changes

Move PinResponse to before usage
Move simplified firstAuth() to authorize() and move interactivity back into firstAuth()
Use PinResponse pointers
Rename GetPin to Authorize
More documentation for Authorize

rspier commented:
Thanks for the changes.  I think authCache is all that's left to deal with.

Commit:
Rename all occurences of "authCache" to "cacheFile"

rspier commented:
Merged.  Thanks!

tedpearson commented:
Thank you for your prompt review and feedback!
